### PR TITLE
Edit slot  in Coordinator controller

### PR DESCRIPTION
**Pull request by mariiasirenko2**

### CODE REVIEW

**Comment by knurff** [comment]
|File|Coverage [6.75%]|:x:|
|:-|:-:|:-:|
|[DateUtils.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FDateUtils.java)|66.04%|:green_apple:|
|[InterviewerSlot.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FInterviewerSlot.java)|21.48%|:green_apple:|
|[User.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FUser.java)|10.78%|:green_apple:|
|[DayOfWeekConvertor.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FDayOfWeekConvertor.java)|0%|:x:|
|[MappingUtils.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FMappingUtils.java)|0%|:x:|
|[InterviewerSlotDto.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2Fdto%2FInterviewerSlotDto.java)|0%|:x:|
|[InterviewerService.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2FInterviewerService.java)|0%|:x:|
|[CoordinatorService.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2FCoordinatorService.java)|0%|:x:|
|[CoordinatorController.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2FCoordinatorController.java)|0%|:x:|
|[InterviewerSlotsDtoCreator.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2Fdto%2Fcreator%2FInterviewerSlotsDtoCreator.java)|0%|:x:|
|[TimePeriodValidator.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/2b83d15dc77071164d8e9e53925b0f995a7fd97d/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2Fvalidator%2FTimePeriodValidator.java)|0%|:x:|

|Total Project Coverage|15.92%|:green_apple:|
|:-|:-:|:-:|

File: src/main/java/com/intellias/intellistart/interviewplanning/controller/dto/InterviewerSlotDto.java [hunk 1]
```diff
@@ -1,18 +1,24 @@
...
 import lombok.NoArgsConstructor;
+import lombok.Setter;
 
 /**
  * InterviewerSlot dto.
  */
 @Data
 @Builder
+@Getter
+@Setter
```
Comment (andriimanzar) [inline src/main/java/com/intellias/intellistart/interviewplanning/controller/dto/InterviewerSlotDto.java:21]: Аннотації Getter та Setter можна не використовувати, адже клас вже помічений аннотацією Data

File: src/main/java/com/intellias/intellistart/interviewplanning/service/InterviewerService.java [hunk 2]
```diff
@@ -72,4 +75,62 @@ public User getInterviewerOrThrowException(Long interviewerId) {
...
+
+    if (isSlotOverlapping(interviewerSlots, slot)) {
+      throw new SlotIsOverlappingException("Slot already exist");
+    }
+  }
+
+  private boolean isSlotOverlapping(List<InterviewerSlot> interviewerSlots, InterviewerSlot slot) {
+    return interviewerSlots.stream()
+        .filter(anotherSlot -> !anotherSlot.getId().equals(slot.getId()))
+        .filter(anotherSlot -> anotherSlot.getDate().equals(slot.getDate()))
```
Comment (andriimanzar) [inline src/main/java/com/intellias/intellistart/interviewplanning/service/InterviewerService.java:132]: Фільтри варто було б винести в окремий метод

File: src/main/java/com/intellias/intellistart/interviewplanning/service/CoordinatorService.java [hunk 3]
```diff
@@ -1,32 +1,68 @@
...
+      InterviewerSlot newSlot) {
+
+    User user = findById(interviewerId);
+    InterviewerSlot oldSlot = interviewerService.findSlotByIdAndInterviewerId(interviewerSlotId,
+        interviewerId);
+
+    if (!oldSlot.getBookings().isEmpty()) {
+      throw new SlotContainsBookingsException(
+          "InterviewerSlot id = " + interviewerSlotId + " has bookings");
+    }
```
Comment (andriimanzar) [inline src/main/java/com/intellias/intellistart/interviewplanning/service/CoordinatorService.java:56]: Логіку знаходження oldSlot та перевірку його букінгів варто винести в окремий метод

File: src/main/java/com/intellias/intellistart/interviewplanning/service/CoordinatorService.java [hunk 4]
```diff
@@ -1,32 +1,68 @@
...
+
+    newSlot.setId(interviewerSlotId);
+    newSlot.setInterviewerId(user);
+
+    DateUtils.checkDateIsInFuture(newSlot.getDate());
+    TimePeriodValidator.checkTimePeriod(newSlot.getPeriod());
+    interviewerService.checkSlotOverlapping(newSlot);
+
```
Comment (andriimanzar) [inline src/main/java/com/intellias/intellistart/interviewplanning/service/CoordinatorService.java:64]: Я думаю, тут варто створити хелпер метод validateNewSlot і там перевіряти його валідність

**Comment by knurff** [comment]
|File|Coverage [6.61%]|:x:|
|:-|:-:|:-:|
|[DateUtils.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FDateUtils.java)|66.04%|:green_apple:|
|[InterviewerSlot.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FInterviewerSlot.java)|21.48%|:green_apple:|
|[User.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fmodel%2FUser.java)|10.78%|:green_apple:|
|[DayOfWeekConvertor.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FDayOfWeekConvertor.java)|0%|:x:|
|[MappingUtils.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Futil%2FMappingUtils.java)|0%|:x:|
|[InterviewerSlotDto.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2Fdto%2FInterviewerSlotDto.java)|0%|:x:|
|[InterviewerService.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2FInterviewerService.java)|0%|:x:|
|[CoordinatorService.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2FCoordinatorService.java)|0%|:x:|
|[CoordinatorController.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2FCoordinatorController.java)|0%|:x:|
|[InterviewerSlotsDtoCreator.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fcontroller%2Fdto%2Fcreator%2FInterviewerSlotsDtoCreator.java)|0%|:x:|
|[TimePeriodValidator.java](https://github.com/knurff/intellistart-java-2022-NPE_HUNTERS/blob/21b1e80d14bbc4efe18c5ad50df9ec0a7c19e62c/src%2Fmain%2Fjava%2Fcom%2Fintellias%2Fintellistart%2Finterviewplanning%2Fservice%2Fvalidator%2FTimePeriodValidator.java)|0%|:x:|

|Total Project Coverage|15.77%|:green_apple:|
|:-|:-:|:-:|

**Review by andriimanzar — APPROVED** [review]
Great job!

**Pull request merged**